### DISCUSSION TITLE
How to find Satellite current revolution number or revolutions since epoch?

### DISCUSSION THREAD

**Discussion by DJCstore** [skyfielders/python-skyfield#880] [Q&A]
Is there an easy way to calculate the current Satellite revolution number, or the number of revolutions since epoch?

My code returns the number of revolutions at epoch with model.revnum but I cant seem to find how to get the current revolution number or calculate it.

**Comment by brandon-rhodes** [comment]
I'm not myself aware of a way, but maybe someone else can weigh in. I don't even see an attribute in the SGP4 model that would offer the satellite's period, from which revolution number could be computed manually.

EDIT: Although the official SGP4 implementation that Skyfield uses doesn't report (and probably doesn't even know) the revolution number of a position that it computes for the satellite, it does at least make available the revolution number that's given for the satellite in its TLE:

```
from skyfield import api
sat = api.EarthSatellite(
    '1 25544U 98067A   20114.53491251  .00001443  00000-0  33988-4 0  9995',
    '2 25544  51.6434 258.9578 0001655 179.6828 273.6120 15.49306451223542',
)
print(sat.model.revnum)
```

Result:

```
22354
```

**Reply by brandon-rhodes** [reply]
A fun fact that I ran across: the official SGP4 satellite orbit implementation that Skyfield uses actually does have this line of code in the satellite record structure:

```
  double period_sec; // Period in seconds
```

But it’s never again used in the code! Alas. They must have had such high hopes, but they were dashed.

**Reply by SpinStabilized** [reply]
I would submit this update myself but my life is crazy at the moment. If you want to add it as a property to a satellite object, here is a version of what I did in my code to get the orbit period.

```python
    @property
    def period(self) -> datetime.timedelta:
        """The orbital period of the spacecraft.
        
        """
        period_m: float = 1 / (self.model.no_kozai / (2 * numpy.pi))
        return datetime.timedelta(minutes=period_m)
```

In my code, I have a wrapper `Spacecraft` object that identifies the `skyfield.sgp4lib.EarthSatellite` for the spacecraft as a `model` so in my code, I call `self.model.model.no_kozai`. Confusing but at this point the code is only for me and has not been shared with anyone. I'm just using it to identify when orbits start at the ascending node and to identify ground station contact periods. I took out the extra `model` in the call for my code

**Comment by SpinStabilized** [comment]
I am looking for this myself and this confirms that I am not missing something in the library that provides this information already. I am working with launch TLE so the epoch isn't at the ascending node. I am going to write my own code that will search for an ascending node and then propagate based on the period. It will just look for the latitude to go from negative to positive as a latitude crossing. You can derive the period from the Mean Motion as the Mean Motion is the number of orbits per day so the time per is the reciprocal.

**Comment by DJCstore** [comment]
I'm a' hobby level' Pythoner, so rather struggle to get where I need to get. 
Skyfield is a great library & given that the orbit number is in the TLE at time/date of issue its a shame there isn't a routine in Skyfield. 
I guess its trivial for a competent coder to calculate & I just need to get on it!
I'm trying to automate an antenna tracker having been disappointed with some aspects of available programs such as Gpredict.
I want to be able to set up a batch of satellite passes to auto track and using the orbit number seems a good way to index & catalogue them and reference rather date/time AOS etc. I'd be interested in any code you come up with. :)

**Reply by SpinStabilized** [reply]
I need to clean up my code because I'm not using the vectorized time methods like I should be to speed up calculations but what I do, essentially, is:

1. Get the current epoch, orbit period, and orbit number
2. Shift back by 1 orbit in time (`epoch - orbit_period`)
3. Build a list of positions (using `at()`)
4. Filter the list for where the latitude is less than zero. The last element in this list is the last ascending node before the epoch. Should be right around the epoch itself if the epoch was defined at the ascending node.

I then use this to see how many orbital periods have passed since the last ascending node when calculating the orbit number at a specific time. It is probably buggy as a general algorithm and needs to be cleaned up but works for my specific condition right now where the epoch is defined at the tip-off from the rocket second stage and need to find when orbit 0 would have actually started.

Hope this helps. Don't have that code handy at the moment.

**Reply by DJCstore** [reply]
Thanks for the info.
I've now got my antenna tracker interface running really good, so can now work on the batching up the forthcoming passes into some code. Hopefully with your suggestions/info it gives me a good steer to get to where I need to go.
Thanks
